Support for dynamic script dependencies in elements (#8681)
PrairieLearn/PrairieLearn: docs/elementExtensions.md

    "clientFilesCourseScripts": ["Client files course JavaScripts"],
    "extensionStyles": ["Styles in the extension directory"],
    "extensionScripts": ["JavaScripts in the extension directory"]
  },
  "dynamicDependencies": {
    "nodeModulesScripts": {"Node modules JavaScripts"},
    "clientFilesCourseScripts": {"Client files course JavaScripts"},

>     "extensionScripts": {"JavaScripts in the extension directory"}

Reviewer: FWIW, this isn't valid JSON. Can we make this example more realistic?

### Extension Dependencies

> Similar to how questions and elements may require client-side assets, extensions may also require client-side JavaScript and CSS. The different properties are summarized here. Note that script dependencies may be set as either static or dynamic dependencies, while styles may only be set as static dependencies.

Reviewer: Consider linking to the `devElements` page, which contains valuable information about how to actually use dynamic dependencies.

| Property                   | Description                                                                                                                                     |
| -------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- |
| `nodeModulesStyles`        | The styles required by this extension, relative to `[PrairieLearn directory]/node_modules`.                                                     |
| `nodeModulesScripts`       | The scripts required by this extension, relative to `[PrairieLearn directory]/node_modules`.                                                    |

> | `extensionStyles`          | The styles required by this element relative to the extension's directory,`[course directory]/elementExtensions/element-name/extension-name`.   |

Reviewer: ```suggestion
| `extensionStyles`          | The styles required by this element relative to the extension's directory, `[course directory]/elementExtensions/element-name/extension-name`.  |
```

| `extensionScripts`         | The scripts required by this element relative to the extension's directory, `[course directory]/elementExtensions/element-name/extension-name`. |
| `clientFilesCourseStyles`  | The styles required by this extension relative to `[course directory]/clientFilesCourse`.                                                       |
| `clientFilesCourseScripts` | The scripts required by this extension relative to `[course directory]/clientFilesCourse`.                                                      |

> Note that all client-side extension assets are always loaded, regardless of whether their Python controller was loaded or not. As such, it is recommended that, when suitable, extensions make use of dynamic dependencies that load scripts only as necessary, based on the context of the element.

Reviewer: This first sentence doesn't quite mesh with the second. Maybe this is clearer?

```suggestion
Note that any element extension assets declared in `dependencies` will always be loaded, regardless of whether their Python controller was loaded or not. As such, it is recommended that, when suitable, extensions make use of `dynamicDependencies` to load scripts only when necessary, based on the context/usage of the element.
```
